runtime/src/main/java/org/corfudb/protocols/wireprotocol/LogData.java

```diff
@@ -72,29 +73,38 @@ public Object getPayload(CorfuRuntime runtime) {
                     if (data == null) {
                         this.payload.set(null);
                     } else {
-                            ByteBuf copyBuf = Unpooled.wrappedBuffer(data);
-                            final Object actualValue;
-                            try {
-                                actualValue =
-                                        Serializers.CORFU.deserialize(copyBuf, runtime);
-                            } catch (Throwable throwable) {
-                                log.error("Exception caught at address {}, {}, {}",
-                                        getGlobalAddress(), getStreams(), getType());
-                                log.error("Raw data buffer {}",
-                                        copyBuf.resetReaderIndex().toString(Charset.defaultCharset()));
-                                copyBuf.release();
-                                data = null;
-                                throw throwable;
-                            }
-                            if (actualValue instanceof LogEntry) {
-                                ((LogEntry) actualValue).setGlobalAddress(getGlobalAddress());
-                                ((LogEntry) actualValue).setRuntime(runtime);
-                            }
-                            value = actualValue == null ? this.payload : actualValue;
-                            this.payload.set(value);
+                        ByteBuf copyBuf = Unpooled.wrappedBuffer(data);
+                        if (hasPayloadCodec()) {
+                            // if the payload has a codec we need to decode it before deserialization
+                            ByteBuf temp = ICorfuPayload.fromBuffer(data, ByteBuf.class);
+                            byte[] temp2 = new byte[temp.readableBytes()];
+                            temp.readBytes(temp2);
+                            copyBuf = Unpooled.wrappedBuffer(getPayloadCodecType()
+                                    .getInstance().decompress(ByteBuffer.wrap(temp2)));
+                        }
+
+                        final Object actualValue;
+                        try {
+                            actualValue =
+                                    Serializers.CORFU.deserialize(copyBuf, runtime);
+                        } catch (Throwable throwable) {
+                            log.error("Exception caught at address {}, {}, {}",
+                                    getGlobalAddress(), getStreams(), getType());
+                            log.error("Raw data buffer {}",
+                                    copyBuf.resetReaderIndex().toString(Charset.defaultCharset()));
                             copyBuf.release();
-                            lastKnownSize = data.length;
                             data = null;
+                            throw throwable;
+                        }
```

Comment:
You can use `finally` here to consolidate the logic.

Reply:
Done.